cmd/controller-manager/app/controller-manager.go

```diff
@@ -117,6 +117,17 @@ func Run(opts *options.Options) error {
 		klog.Fatalf("Invalid Feature Gate: %v", err)
 	}
 
+	// for in-cluster deployment set the namespace associated
+	// with the service account token
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
```

Comment:
We already have a valid config at this point, and the logic for discovering the namespace of the service account doesn't depend on the config at all.  It's just [5 lines of go code](https://github.com/kubernetes-sigs/federation-v2/blob/master/vendor/k8s.io/client-go/tools/clientcmd/client_config.go#L521).  I'd suggest creating a local function with this code or just adding it inline:

```golang
// Fall back to the namespace associated with the service account token, if available
if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
  if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
    return ns, false, nil
  }
}
```

If the federation namespace is not specified and cannot be discovered via this method, the controller manager should log a fatal error indicating that the federation namespace needs to be manually set via `--federation-config` if the controller manager is not running inside a cluster.

Reply:
Ack. I will make the necessary changes in the code as per your suggestions.